packages/react-menu/src/components/Menu/useMenu.tsx

```diff
@@ -59,6 +72,86 @@ export const useMenu = (props: MenuProps, ref: React.Ref<HTMLElement>, defaultPr
     },
     [setOpen],
   );
+  state.on.length === 0 && state.on.push('click');
+
+  useMenuPopup(state);
+  useOnClickOutside({ element: document, refs: [state.menuPopupRef, triggerRef], callback: () => setOpen(false) });
+  useOnEscape({ element: document, callback: () => setOpen(false) });
+  return state;
+};
+
+const useMenuPopup = (state: MenuState) => {
+  // TODO use Popper for the popup ref
+  const menuPopupRef = React.useRef<HTMLElement>(null) as React.MutableRefObject<HTMLElement>;
+  state.menuPopupRef = menuPopupRef;
+  state.menuPopup.children = (Component, originalProps) => {
+    const newProps = { 'aria-labelledBy': state.triggerId, ...originalProps };
+
+    if (state.on.includes('hover')) {
+      newProps.onMouseEnter = (e: React.MouseEvent<HTMLElement>) => {
+        state.setOpen(true);
+        originalProps?.onMouseEnter?.(e);
+      };
+      newProps.onMouseLeave = (e: React.MouseEvent<HTMLElement>) => {
+        state.setOpen(false);
+        originalProps?.onMouseLeave?.(e);
+      };
+    }
+
+    newProps.onBlur = (e: React.FocusEvent<HTMLElement>) => {
+      state.setOpen(false);
+      originalProps?.onBlur?.(e);
+    };
+
+    return (
+      // Shorthand render funct types have issues
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
+      <Component {...newProps} ref={state.menuPopupRef}>
+        {state.menuList}
+      </Component>
+    );
+  };
 
   return state;
 };
+
+// TODO handle nested menus
+const useOnClickOutside = (options: {
+  element?: Node | Window | Document;
+  refs: React.MutableRefObject<HTMLElement>[];
+  callback: (ev: MouseEvent) => void;
+}) => {
+  const { refs, callback, element = document } = options;
+
+  React.useEffect(() => {
+    const listener = (ev: MouseEvent) => {
+      const isOutside = !refs.some(ref => elementContains(ref.current, ev.target as HTMLElement));
+      if (isOutside) {
+        callback(ev);
+      }
+    };
+
+    element?.addEventListener('click', listener);
+
+    return () => element?.removeEventListener('click', listener);
+  }, [refs, callback, element]);
```

Comment:
This should use `useEventCallback` as closures will destroy memoization:

```tsx
useOnClickOutside({ 
  element: document,
  refs: [state.menuPopupRef, triggerRef], // ⚠️ a new reference as an array will be created on each render
  callback: () => setOpen(false) // ⚠ a new reference as closures are unstable
});
```

A proposed change:

```tsx
const useOnClickOutside = (options: {
  element?: Node | Window | Document;
  refs: React.MutableRefObject<HTMLElement>[];
  callback: (ev: MouseEvent) => void;
}) => {
  const { refs, callback, element = document } = options;
  const listener = useEventCallback((ev: MouseEvent) => {
    const isOutside = !refs.some(ref => elementContains(ref.current, ev.target as HTMLElement));

    if (isOutside) {
      callback(ev);
    }
  });

  React.useEffect(() => {
    element?.addEventListener('click', listener);

    return () => element?.removeEventListener('click', listener);
  }, [element, listener]);
};
```